AraEvent/AraEventCalibrator.cxx

```diff
@@ -1458,11 +1491,159 @@ Bool_t AraEventCalibrator::TrimFirstBlock(UsefulAtriStationEvent *theEvent, std:
                 sampleList->at(chanId).erase(sampleList->at(chanId).begin(),sampleList->at(chanId).begin()+first_block_len); ///< delete the samples in the first block
+                blockList->at(chanId).erase(blockList->at(chanId).begin(),blockList->at(chanId).begin()+first_block_len); ///< delete the samples in the first block
             }
         }
     } return !hasTooFewBlocks;
 }
 
+//! Correct block offsets
+/*!
+    \param theEvent the useful atri event pointer
+    \param voltMapIt the iterator referring to the voltage elements in the 2d map container
+    \param timeMapIt the iterator referring to the time elements in the 2d map container
+    \param blockList the pointer of block number of each sample
+    \return void
+*/
+void AraEventCalibrator::CorrectBlockOffset(UsefulAtriStationEvent *theEvent, std::map< Int_t, std::vector <Double_t> >::iterator &voltMapIt, std::map< Int_t, std::vector <Double_t> >::iterator &timeMapIt, std::vector<std::vector<int> > *blockList)
+{
+    for(int dda=0; dda<DDA_PER_ATRI; dda++) {
+        
+        for(Int_t chan=0; chan<RFCHAN_PER_DDA/2; chan++) { // only need to correct the used channels
+            Int_t chanId=chan+RFCHAN_PER_DDA*dda;
+            voltMapIt=theEvent->fVolts.find(chanId);
+            timeMapIt=theEvent->fTimes.find(chanId);
+            std::vector<int>& blkList = blockList->at(chanId);
+            if(voltMapIt!=theEvent->fVolts.end()) {
+                int total_samples = std::distance(voltMapIt->second.begin(), voltMapIt->second.end());
+                if(total_samples != int(blkList.size()))
+                    throw std::runtime_error("Block list length not equal to number of samples!");
+
+                // create graph of calibrated trace
+                std::vector<double> tmpT(timeMapIt->second.begin(), timeMapIt->second.end());
+                std::vector<double> tmpV(voltMapIt->second.begin(), voltMapIt->second.end());
+
+                
+                // find block edges
+                std::vector<double> blockStarts; // times of the block starts
+                std::vector<double> blockEnds; // times of the block ends
+                blockStarts.push_back(tmpT[0]);
+                for(int i = 1; i < total_samples; ++i) {
+                    if(blkList[i] != blkList[i-1]) {
+                        blockStarts.push_back(tmpT[i]);
+                        blockEnds.push_back(tmpT[i]);
+                    }
+                }
+                blockEnds.push_back(tmpT[total_samples-1]);
+                
+                const int num_block = int(blockStarts.size());
+                std::vector<double> dT(num_block); // width of block
+                for(int i = 0; i < num_block; ++i) 
+                    dT[i] = blockEnds[i] - blockStarts[i]; 
+
+                // interpolate onto regular 0.5 ns timestep
+                TGraph* gr = new TGraph(total_samples, &tmpT[0], &tmpV[0]); 
+                const double dt = 0.5; // ns
```

Comment:
why 0.5 ns?

Reply:
This is arbitrary, I just do it to get the waveform onto a constant timestep for the FFT.

Reply:
Is it sensitive to the choice at all?

Reply:
I'm not sure? I guess in the same way any FFT of a waveform we've interpolated is? I don't think this would be particularly sensitive since we are only looking at the spectrum below 50 MHz in the end.

Reply:
Ahh, okay, I forgot about the really vigorous bandpass. That seems okay then.